docs/requirements.md

```diff
@@ -0,0 +1,149 @@
+
+Locales manifest in two primary ways. First, they affect selection of fonts from a fallback list, especially Han unification. Second, they need to be plumbed to the shaping engine so OpenType [`locl`] features can be selected.
+
+Note that to really do this properly requires a locale *list,* not just a single locale as in some older interfaces. An example would be two users with en_US as their primary locale, but with ja_JP and zh_CN as their secondary locales. In this case, the secondary locale would be responsible for Han unification. In short, Han unification is controlled by the first CJK entry in the locale list. But Han unification is only the most visible example; in general, localization needs to be done from the first relevant entry, and the logic for what's relevant is complex and font-dependent.
+
+## Performance
+
+Text layout can be slow, and especially as more of the rendering pipeline is moving to the GPU, it can be the limiting factor. A large part of the design goal is performance, and here are some mechanisms to try to achieve that:
+
+* A layout cache. One subtlety is that it should be designed for multithreaded use, so a `Mutex<HashMap>` may replaced at some point with a concurrent hash map.
+
+* A fast representation of styles. Since style information can be expensive, both for storage and for hash-table lookup, we might design an arena, indexed by integer. Alternatively, we might use a small struct that covers the most common cases, with an optional `Box` for fancier style parameters such as OpenType features. In any case, locale is a large part of the puzzle.
+
+* Possibly fast paths for monospace fonts, including CJK. These have to be carefully designed to preserve semantics (unlike the "simple path" in Blink before it was removed).
+
+## Some open questions
+
+* Do shaping driven itemization, or use coverage data extracted from fonts (as in Minikin)?
+
+* Is layout at this level size-independent, or can it take into account pixel size dependent metrics (see [font-kit#34](https://github.com/pcwalton/font-kit/issues/34)).
+
+* Do we have a dependency on font-kit in all cases, or only cases where we're not using the platform layout? The font-kit dependency seems pretty heavyweight; more empirical measurement is necessary here.
+
+## Vertical text
+
+Vertical text is now part of the requirements for fully correct Web text rendering, thanks to [CSS writing modes]. Thus, vertical text is part of the *design* for skribo. I'd like more information about whether its implementation should be a priority.
+
+Some of the groundwork can be done just by using proper types (vector rather than scalar for advance, an enum rather than bool for text direction).
```

Comment:
Apparently in DirectWrite one passes a flag to rotate (some) glyphs 90° counter-clockwise, then does horizontal layout, then rotates entire lines 90° clockwise when rendering: https://docs.microsoft.com/en-us/windows/desktop/directwrite/vertical-text